### PR TITLE
Improve mptcpd ID manager hashing.

### PR DESCRIPTION
**Pull request by ossama-othman**
Replace the existing trivial `mptcp_idm` key hashing with a hash based on the much better public domain [MurmurHash3](https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp) implementation found in the newly added `mptcpd_murmur_hash3()` function.

Changes in mptcpd's version of MurmurHash3 include the folliowing:

* 128 bit hash functions were removed.  They are not needed by mptcpd since the ELL l_hashmap expects hash values of type `unsigned int.`
* The `MurmurHash3_x86_32()` function was renamed to `mptcpd_murmur_hash3()`.
* The hash value is returned as an unsigned int return value instead of a function "_out_" parameter of type `void*`.
* The only compiler-specific support left in place is for gcc and clang.
* Declare `inline` functions as `static inline`.
* The coding style was modified to conform to the [mptcpd style](https://github.com/intel/mptcpd/blob/master/CONTRIBUTING.md).
* gcc "implicit fallthrough" warnings for switch statement were silenced by leveraging the appropriate [gcc statement attribute](https://gcc.gnu.org/onlinedocs/gcc/Statement-Attributes.html).

### CODE REVIEW

File: lib/id_manager.c [hunk 1]
```diff
@@ -203,6 +222,34 @@ struct mptcpd_idm *mptcpd_idm_create(void)
...
+                                           errmsg,
+                                           L_ARRAY_SIZE(errmsg));
+
+                        l_error("Unable to generate ID manager "
+                                "hash seed: %s",
+                                r == 0 ? errmsg : "<unknown error>");
+                }
+#else
+                // Fallback on current time as seed (not desirable).
+                _idm_hash_seed = (uint32_t) time(NULL);
```
Comment (mjmartineau) [inline lib/id_manager.c:249]: I'd much prefer fallback to `l_getrandom_uint32()`. That function even has additional fallback to `random()` if the syscall fails.

File: lib/id_manager.c [hunk 2]
```diff
@@ -203,6 +222,34 @@ struct mptcpd_idm *mptcpd_idm_create(void)
...
+                                  sizeof(_idm_hash_seed),
+                                  GRND_NONBLOCK);
+
+                if (len == -1) {  // unlikely
+                        char errmsg[80];
+                        int const r =
+                                strerror_r(errno,
+                                           errmsg,
+                                           L_ARRAY_SIZE(errmsg));
+
```
Comment (mjmartineau) [inline lib/id_manager.c:242]: Could also fail over to `l_getrandom_uint32()` here and make it a warning instead of error.

File: lib/id_manager.c [hunk 3]
```diff
@@ -203,6 +222,34 @@ struct mptcpd_idm *mptcpd_idm_create(void)
...
+                        l_error("Unable to generate ID manager "
+                                "hash seed: %s",
+                                r == 0 ? errmsg : "<unknown error>");
+                }
+#else
+                // Fallback on current time as seed (not desirable).
+                _idm_hash_seed = (uint32_t) time(NULL);
+#endif  // HAVE_GETAUXVAL
```
Comment (mjmartineau) [inline lib/id_manager.c:250]: Stale comment :)

**Review by mjmartineau — CHANGES_REQUESTED** [review]
Commit structure looks good, thanks! Just a couple of seed init comments.

File: lib/id_manager.c (hunk 1 above)
Comment (ossama-othman) [reply lib/id_manager.c:249]: I'd rather drop the fallback completely, and stick with `l_getrandom_uint32()`.

File: lib/id_manager.c (hunk 2 above)
Comment (ossama-othman) [reply lib/id_manager.c:242]: Let's just use `l_getrandom_uint32()`.

**Comment by ossama-othman** [comment]
Proposed diff (posted here to minimize commits since we won't be squashing this PR):
```diff
diff --git a/configure.ac b/configure.ac
index 723b519..6b7c575 100644
--- a/configure.ac
+++ b/configure.ac
@@ -267,11 +267,6 @@ AC_CHECK_FUNC([l_hashmap_replace],
                          [ELL has l_hashmap_replace()])])
 LIBS=$mptcpd_save_libs
 
-AC_CHECK_FUNC([getrandom],
-              [AC_DEFINE([HAVE_GETRANDOM],
-                         [],
-                         [C library has getrandom()])])
-
 # ---------------------------------------------------------------
 # Enable additional C compiler warnings.  We do this after all
 # Autoconf tests have been run since not all autoconf macros are
diff --git a/lib/id_manager.c b/lib/id_manager.c
index dd027af..fb9bb8c 100644
--- a/lib/id_manager.c
+++ b/lib/id_manager.c
@@ -21,16 +21,13 @@
 #include <sys/socket.h>
 #include <netinet/in.h>
 
-#ifdef HAVE_GETRANDOM
-# include <sys/random.h>
-#endif // HAVE_GETRANDOM
-
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wpedantic"
 #include <ell/hashmap.h>
 #include <ell/uintset.h>
 #include <ell/util.h>
 #include <ell/log.h>
+#include <ell/random.h>
 #pragma GCC diagnostic pop
 
 #include <mptcpd/private/murmur_hash.h>
@@ -222,33 +219,8 @@ struct mptcpd_idm *mptcpd_idm_create(void)
                 idm = NULL;
         }
 
-        if (_idm_hash_seed == 0) {
-#ifdef HAVE_GETRANDOM
-                /*
-                  Randomize the hash seed value with data from the
-                  kernel random number generator.
-                */
-                ssize_t const len =
-                        getrandom(&_idm_hash_seed,
-                                  sizeof(_idm_hash_seed),
-                                  GRND_NONBLOCK);
-
-                if (len == -1) {  // unlikely
-                        char errmsg[80];
-                        int const r =
-                                strerror_r(errno,
-                                           errmsg,
-                                           L_ARRAY_SIZE(errmsg));
-
-                        l_error("Unable to generate ID manager "
-                                "hash seed: %s",
-                                r == 0 ? errmsg : "<unknown error>");
-                }
-#else
-                // Fallback on current time as seed (not desirable).
-                _idm_hash_seed = (uint32_t) time(NULL);
-#endif  // HAVE_GETAUXVAL
-        }
+        if (_idm_hash_seed == 0)
+                _idm_hash_seed = l_getrandom_uint32();
 
         return idm;
 }
```

**Comment by mjmartineau** [comment]
That proposed diff switching to `l_getrandom_uint32()` looks good to me.

**Review by mjmartineau — APPROVED** [review]
Looks good to me.

**Pull request merged**